sdks/java/core/src/test/java/org/apache/beam/sdk/io/AvroIOTransformTest.java

```diff
@@ -293,29 +312,33 @@ public void testRead() throws Exception {
     @Parameterized.Parameter(1)
     public String testAlias;
 
-    private <T> void runTestWrite(final AvroIO.Write<T> writeBuilder)
-        throws Exception {
+    @Parameterized.Parameter(2)
+    public Object[] inputData;
+
+    @Parameterized.Parameter(3)
+    public Boolean generic;
 
+    private <T> void runTestWrite(final AvroIO.Write<T> writeBuilder, T[] inputData,
+        boolean generic) throws Exception {
       final File avroFile = tmpFolder.newFile("file.avro");
-      final AvroGeneratedUser[] users = generateAvroObjects();
       final AvroIO.Write<T> write = writeBuilder.to(avroFile.getPath());
 
       @SuppressWarnings("unchecked") final
       PCollection<T> input =
-          pipeline.apply(Create.of(Arrays.asList((T[]) users))
+          pipeline.apply(Create.of(Arrays.asList(inputData))
                                .withCoder((Coder<T>) AvroCoder.of(AvroGeneratedUser.class)));
       input.apply(write.withoutSharding());
 
       pipeline.run();
 
       assertEquals(WRITE_TRANSFORM_NAME, write.getName());
-      assertThat(readAvroFile(avroFile), containsInAnyOrder(users));
+      assertThat(readAvroFile(avroFile, generic), containsInAnyOrder((Object[]) inputData));
```

Comment:
Wonder if this would be simpler to do as a "write-then-read" pipeline: instead of using Avro API to read the results, use AvroIO.read().

Reply:
Well, I'm used to avoid using IO read verb when I test write verb in case there is an error with the read, it would be misplaced (showing in the write test). Generally speaking I like to have tests that test single features as much as possible. This is why I stayed (it was done this way before) with reading through the Avro api.

Reply:
I think "write, then read" is not prone to the issue you mention. Specifically, we do the following:

- Generate a set R of test records
- Write R to place X
- Read them back from X
- Assert that read result == R

If there's an error with the read implementation, it may produce a failure even if write was correct (mistakenly return something != R) - however, I don't think it can mask a failure when write is incorrect: I have a very hard time imagining how a broken read() implementation can return data that doesn't equal what write() wrote, but equals R by accident. We have quite a few tests in this style; and some testing scenarios are hugely easier to test in this way (eg. see https://github.com/apache/beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/io/AvroIOTest.java#L303).

I'm fine keeping the code as-is in this PR if you'd like, because it was indeed already done this way and it's unrelated to the purpose of this PR, but I think then I'll be tempted to make this change myself in a separate PR :)

Reply:
Indeed, IMHO I prefer that this change is done in another PR because unrelated to genericity.